sdk/monitor/azure-monitor-opentelemetry-exporter/azure/monitor/opentelemetry/exporter/_quickpulse/_types.py

```diff
@@ -0,0 +1,173 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+from dataclasses import dataclass
+from typing import Dict
+
+from opentelemetry.sdk._logs import LogRecord
+from opentelemetry.sdk.trace import ReadableSpan
+from opentelemetry.semconv.trace import SpanAttributes
+from opentelemetry.trace import SpanKind
+from opentelemetry.util.types import Attributes
+
+from azure.monitor.opentelemetry.exporter.export.trace import _utils as trace_utils
+
+@dataclass
+class _TelemetryData:
+    custom_dimensions: Dict[str, str]
+
+    @staticmethod
+    def _from_span(span: ReadableSpan):
+        if span.kind in (SpanKind.SERVER, SpanKind.CONSUMER):
+            return _RequestData._from_span(span)
+        else:
+            return _DependencyData._from_span(span)
+        
+    @staticmethod
+    def _from_log_record(log_record: LogRecord):
+        exc_type = log_data.log_record.attributes.get(SpanAttributes.EXCEPTION_TYPE)  # type: ignore
+        exc_message = log_data.log_record.attributes.get(SpanAttributes.EXCEPTION_MESSAGE)  # type: ignore
+        if exc_type is not None or exc_message is not None:
+            return _ExceptionData._from_log_record(log_record)
+        else:
+            return _TraceData._from_log_record(log_record)
+
+
+@dataclass
+class _RequestData(_TelemetryData):
+    duration: float
+    success: bool
+    name: str
+    response_code: int
+    url: str
+
+    @staticmethod
+    def _from_span(span: ReadableSpan):
+    # Logic should match that of exporter to Breeze
+        url = ""
+        duration_ms = 0
+        response_code = 0
+        success = True
+        attributes = {}
+        if span.end_time and span.start_time:
+            duration_ms = (span.end_time - span.start_time) / 1e9  # type: ignore
+        if span.attributes:
+            attributes = span.attributes
+            url = span.attributes.get(SpanAttributes.HTTP_URL, "")
```

Comment:
This way of fetching url looks simpler than what node.js does - I'm curious at a high level if the way of fetching url is speced somewhere or agreed upon. In the case where the HTTP_URL isn't defined, node.js takes some extra steps to build some url by looking at http scheme / target / host / peer port, name, ip etc.

Reply:
Yes you correct, Python does this too for Request data types for breeze exporter. I need to refactor out this logic from the breeze exporter so ive put a TODO for now.